// vfs-ssh/slave.js
var execFile = require("child_process").execFile;

module.exports = function (config, callback) {

    process.title = "vfs-worker " + JSON.stringify(config);
    var Worker = require('vfs-socket/worker').Worker;
    var vfsLocal = require('vfs-local')(config);
    
    var worker = new Worker(vfsLocal);
    worker.connect([process.stdin, process.stdout], callback);
    worker.on("disconnect", function () {
        process.removeListener("SIGINT", intHandler);
        process.removeListener("SIGTERM", termHandler);
        process.removeListener("uncaughtException", exceptionHandler);
        
        childrenOfPid(process.pid, function(err, pidlist){
            if (err) {
                console.error(err.stack);
                process.exit();
            }
            pidlist.forEach(function (pid) {
                try {
                    process.kill(pid, "SIGKILL");
                } catch(e) {
                    // kill may throw if the pid does not exist.
                }
            });
        });
    });

    process.on("SIGINT", intHandler);
    process.on("SIGTERM", termHandler);
    process.on("uncaughtException", exceptionHandler);

    function intHandler() {
        cleanup(new Error("Received unexpected signal: SIGINT"), 1);
    }
    function termHandler() {
        cleanup(new Error("Received unexpected signal: SIGTERM"), 2);
    }
    function exceptionHandler(err) {
        cleanup(err, 3);
    }
    
    function cleanup(err, code) {
        console.log(process.pid, "Unhandled exception", (err.message || err).toString(), err.stack);
        
        // give some time to send last packages
        setTimeout(function() {
            worker.disconnect(err);
            
            // give some time to close connection
            setTimeout(function() {
                process.exit(code);
            }, 100);
        }, 200);
    }
};

function childrenOfPid(pid, callback) {
    execFile("ps", {args: ["-A", "-oppid,pid"]}, function(err, stdout) {
        if (err)
            return callback(err);

        var parents = {};
        stdout.split("\n").slice(1).forEach(function(line) {
            var col = line.trim().split(/\s+/g);
            (parents[col[0]] || (parents[col[0]] = [])).push(col[1]);
        });

        function search(roots) {
            var res = roots.concat();
            for (var c, i = 0; i < roots.length; i++) {
                if ((c = parents[roots[i]]) && c.length)
                    res.unshift.apply(res, search(c));
            }
            return res;
        }
        callback(null, search([pid]));
    });
}
